// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is govered by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

syntax = "proto3";


package monorail;

import "api/api_proto/common.proto";
import "api/api_proto/project_objects.proto";


service Projects {
  rpc ListProjects (ListProjectsRequest) returns (ListProjectsResponse) {}
  rpc GetConfig (GetConfigRequest) returns (Config) {}
  rpc GetCustomPermissions (GetCustomPermissionsRequest) returns (GetCustomPermissionsResponse) {}
  rpc GetVisibleMembers (GetVisibleMembersRequest) returns (GetVisibleMembersResponse) {}
  rpc GetLabelOptions (GetLabelOptionsRequest) returns (GetLabelOptionsResponse) {}
  rpc ListStatuses (ListStatusesRequest) returns (ListStatusesResponse) {}
  rpc ListComponents (ListComponentsRequest) returns (ListComponentsResponse) {}
  rpc ListFields (ListFieldsRequest) returns (ListFieldsResponse) {}
  rpc GetProjectStarCount (GetProjectStarCountRequest) returns (GetProjectStarCountResponse) {}
  rpc StarProject (StarProjectRequest) returns (StarProjectResponse) {}
  rpc GetUserProjects (GetUserProjectsRequest) returns (GetUserProjectsResponse) {}
  rpc CheckProjectName (CheckProjectNameRequest) returns (CheckProjectNameResponse) {}
}


message ListProjectsRequest {
  int32 page_size = 1;
  string page_token = 2;
}


message ListProjectsResponse {
  repeated Project projects = 1;
  string next_page_token = 2;
}


// Next available tag: 3
message GetConfigRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 3
message GetCustomPermissionsRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 2
message GetCustomPermissionsResponse {
  repeated string permissions = 1;
}


// Next available tag: 3
message GetVisibleMembersRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 3
message GetVisibleMembersResponse {
  repeated UserRef user_refs = 1;
  repeated UserRef group_refs = 2;
}


// Next available tag: 3
message GetLabelOptionsRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 3
message GetLabelOptionsResponse {
  repeated LabelDef label_options = 1;
  repeated string exclusive_label_prefixes = 2;
}


// Next available tag: 3
message ListStatusesRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 4
message ListStatusesResponse {
  repeated StatusDef status_defs = 1;
  repeated StatusRef statuses_offer_merge = 2;
  bool restrict_to_known = 3;
}


// Next available tag: 4
message ListComponentsRequest {
  RequestTrace trace = 1;
  string project_name = 2;
  bool include_admin_info = 3;
}


// Next available tag: 2
message ListComponentsResponse {
  repeated ComponentDef component_defs = 1;
}


// Next available tag: 5
message ListFieldsRequest {
  RequestTrace trace = 1;
  string project_name = 2;
  bool include_admin_info = 3;
  bool include_user_choices = 4;
}


// Next available tag: 2
message ListFieldsResponse {
  repeated FieldDef field_defs = 1;
}


// Next available tag: 3
message GetProjectStarCountRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 2
message GetProjectStarCountResponse {
  uint32 star_count = 1;
}


// Next available tag: 3
message StarProjectRequest {
  RequestTrace trace = 1;
  string project_name = 2;
  bool starred = 3;
}


// Next available tag: 2
message StarProjectResponse {
  uint32 star_count = 1;
}


// Next available tag: 2
message GetUserProjectsRequest {
  RequestTrace trace = 1;
}


// Next available tag: 5
message GetUserProjectsResponse {
  repeated string owner_of = 1;
  repeated string member_of = 2;
  repeated string contributor_to = 3;
  repeated string starred_projects = 4;
}


// Next available tag: 3
message CheckProjectNameRequest {
  RequestTrace trace = 1;
  string project_name = 2;
}


// Next available tag: 1
message CheckProjectNameResponse {
  string error = 1;
}

